### PR TITLE
Pull Neon implementation of Falcon from PQClean

### PR DESCRIPTION
**Pull request by SWilson4**
Pull the ARM Neon implementation of Falcon from PQClean on behalf of @cothan. See https://github.com/PQClean/PQClean/pull/493, https://github.com/PQClean/PQClean/pull/497

* [ ] Does this PR change the input/output behaviour of a cryptographic algorithm (i.e., does it change known answer test values)?  (If so, a version bump will be required from *x.y.z* to *x.(y+1).0*.)
* [ ] Does this PR change the list of algorithms available -- either adding, removing, or renaming? Does this PR otherwise change an API? (If so, PRs in [oqs-provider](https://github.com/open-quantum-safe/oqs-provider), [OQS-OpenSSL](https://github.com/open-quantum-safe/openssl), [OQS-BoringSSL](https://github.com/open-quantum-safe/boringssl), and [OQS-OpenSSH](https://github.com/open-quantum-safe/openssh) will also need to be ready for review and merge by the time this is merged.)

### CODE REVIEW

**Comment by SWilson4** [comment]
@cothan It looks like there are build errors: https://app.circleci.com/pipelines/github/open-quantum-safe/liboqs/2788/workflows/46cbb0ce-3a37-4cfb-ba32-a33f2bced86f/jobs/22376

**Comment by ducnguyen-sb** [comment]
@SWilson4  Thank you, it's type confusion between signed and unsigned integer. Should I make a fix in this PR, or is it better to make a PR in pqclean and update here?

**Comment by SWilson4** [comment]
> @SWilson4 Thank you, it's type confusion between signed and unsigned integer. Should I make a fix in this PR, or is it better to make a PR in pqclean and update here?

I would say it's better to fix it in PQClean and then run `copy_from_upstream` again, but there's been some unrelated activity in PQClean since your initial two commits, and we presumably wouldn't want to pull that in at the same time. So I'm unsure. Thoughts on this @dstebila?

Either way, it's probably best if you can get CI to pass on this branch before making a PR in PQClean, since the CI over there evidently didn't complain about the type confusion. It will save having to iterate through the process again if another build error sneaks in.

**Comment by dstebila** [comment]
> > @SWilson4 Thank you, it's type confusion between signed and unsigned integer. Should I make a fix in this PR, or is it better to make a PR in pqclean and update here?
> 
> I would say it's better to fix it in PQClean and then run `copy_from_upstream` again, but there's been some unrelated activity in PQClean since your initial two commits, and we presumably wouldn't want to pull that in at the same time. So I'm unsure. Thoughts on this @dstebila?

It's not a problem if there are other PQClean updates upstream alongside it, that's sometimes unavoidable based on how our pull-from-upstream process works.

**Comment by cothan** [comment]
Commit https://github.com/open-quantum-safe/liboqs/pull/1547/commits/91065c8a38de045fd4095db44f875cb5ae5cf9c8 makes CI/CD green. 

So I made a PR to fix type confusion in PQClean https://github.com/PQClean/PQClean/pull/506.

**Comment by ducnguyen-sb** [comment]
The PR https://github.com/PQClean/PQClean/pull/506 is merged recently. I think it's alright to proceed.

**Comment by SWilson4** [comment]
I ran `copy_from_upstream` at commit [8e220a87308154d48fdfac40abbb191ac7fce06a](https://github.com/PQClean/PQClean/commit/8e220a87308154d48fdfac40abbb191ac7fce06a) of PQClean and everything is passing.

Besides the Neon implementation of Falcon, there are some minor changes to McEliece from PQClean commit [afcebe31ae6c636f39ce6852b8ee95c3f6e7670d](https://github.com/PQClean/PQClean/commit/afcebe31ae6c636f39ce6852b8ee95c3f6e7670d). A bunch of newlines were deleted, which has made the diff light up like a Christmas tree. Another slight change there is outlined in https://github.com/PQClean/PQClean/pull/503: it makes one of our patches unnecessary.

https://github.com/PQClean/PQClean/pull/501 made another patch unnecessary. I deleted these patches and updated `copy_from_upstream` accordingly.

I also took the liberty of adding `scripts/copy_from_upstream/verify_from_upstream` to the `.gitignore` to prevent me from accidentally committing it after running `./copy_from_upstream -k verify` (oops).

File: docs/algorithms/sig/sphincs.md [hunk 1]
```diff
@@ -66,7 +66,7 @@ Are implementations chosen based on runtime CPU feature detection? **Yes**.
...
 |:---------------------------------:|:-------------------------|:----------------------------|:--------------------------------|:------------------------|:-----------------------------------|:-----------------------------------------------|:---------------------|
 | [Primary Source](#primary-source) | clean                    | All                         | All                             | None                    | True                               | True                                           | False                |
-| [Primary Source](#primary-source) | avx2                     | x86\_64                     | All                             | AVX2                    | True                               | True                                           | False                |
+| [Primary Source](#primary-source) | avx2                     | x86\_64                     | Linux,Darwin                    | AVX2                    | True                               | True                                           | False                |
```
Comment (baentsch) [inline docs/algorithms/sig/sphincs.md:69]: So optimized Sphincs code is no longer supported on Windows? Does anyone know why? Looks like a regression, no?

**Review by baentsch — APPROVED** [review]
Basically LGTM. Code license is clearly documented as Apache2 -- but I didn't find SPDX identifiers in the files. Could those be added going forward, please? This in support of (implementing) #1514 (eventually)?

File: docs/algorithms/sig/sphincs.md (hunk 1 above)
Comment (SWilson4) [reply docs/algorithms/sig/sphincs.md:69]: That's weird... there was no corresponding change in PQClean, where we (presumably?) pulled this data from. I'll take a look at the `copy_from_upstream` script and see if I can figure out what's going on.

Comment (SWilson4) [reply docs/algorithms/sig/sphincs.md:69]: [Link to the relevant file](https://github.com/PQClean/PQClean/blob/8e220a87308154d48fdfac40abbb191ac7fce06a/crypto_sign/sphincs-sha2-192s-simple/META.yml)

Comment (SWilson4) [reply docs/algorithms/sig/sphincs.md:69]: Never mind, this seems to be caused by #1478. Not sure why it didn't get updated in previous pulls from upstream, though.

**Comment by cothan** [comment]
Thanks @baentsch , sure I will update license of Falcon goind forward. 
Sorry I didn't know, I was following the Falcon ref License header example.

File: docs/algorithms/sig/sphincs.md (hunk 1 above)
Comment (baentsch) [reply docs/algorithms/sig/sphincs.md:69]: > Not sure why it didn't get updated in previous pulls from upstream, though.

Don't understand this either. Nor why I didn't catch/check the effect of #1478 to documentation then :-( But then it's now "fixed".

**Comment by baentsch** [comment]
> I will update license of Falcon going forward.

Thanks very much! Please let us know as and when landed in PQClean so we can update suitably.

**Comment by cothan** [comment]
So I made a PR to add SPDX header in https://github.com/PQClean/PQClean/pull/514

**Comment by baentsch** [comment]
> So I made a PR to add SPDX header in [PQClean/PQClean#514](https://github.com/PQClean/PQClean/pull/514)

Much appreciated. But CI isn't exactly happy: How can this be? The diffs are only comments...

**Comment by cothan** [comment]
I totally agree with you. I think the PR changes in Falcon, so it triggers all implementations in Falcon tested. Turn out, AVX2 used defined macros (as it always is from the start) that makes become red.

**Comment by ducnguyen-sb** [comment]
It's been a few weeks https://github.com/PQClean/PQClean/pull/514, I think the other approach can be like this comment: https://github.com/PQClean/PQClean/pull/514#issuecomment-1758039951

**Comment by baentsch** [comment]
> It's been a few weeks [PQClean/PQClean#514](https://github.com/PQClean/PQClean/pull/514), I think the other approach can be like this comment: [PQClean/PQClean#514 (comment)](https://github.com/PQClean/PQClean/pull/514#issuecomment-1758039951)

Agreed. But with the Linux Foundation driven sidelining of independent contributors I in turn sidelined caring about #1514: Should the people paid by entities restricting the breadth of licenses look into this, not people interested in progress of open source.

**Pull request merged**